### ISSUE TITLE
Automated pep8 checks for modified lines

### ISSUE THREAD

**Issue by Nodd** [spyder-ide/spyder#2959]
Following #2370, we should add checks in travis for pep8 correctness, at minimum line feeds and trailing whitespaces.
I know it's possible, but I don't know how !

**Comment by ccordoba12** [comment]
But for that to work, we must start first by adding a new option to Spyder to use a preferred line ending, instead of the operating system one.

Since @goanpeca and me use Spyder to develop Spyder, until we add that option we will continue to introduce line ending inconsistencies :-)

**Comment by goanpeca** [comment]
Well.. I only develop on Kubuntu now... and test on a Windows VM...

**Comment by Nodd** [comment]
Line endings can be managed by git. By default on windows with mysysgit the files on disk have windows-like endings, but when committing they are transformed to linux-like endings, so that in the repo itself there are always linux-like endings.
We just have to state that the line endings should always be linux-like in the repo, and it's up to the developer to correctly setup git (of course we can help with this part and document it).

**Comment by goanpeca** [comment]
@Nodd can you add that somewhere in the wiki ?

**Comment by techtonik** [comment]
https://github.com/spyder-ide/spyder/issues/2424#issuecomment-108285322

**Comment by techtonik** [comment]
Also https://github.com/spyder-ide/spyder/pull/2470

**Comment by techtonik** [comment]
You can merge just two first commits.

**Comment by Nodd** [comment]
@techtonik Sorry I forgot to add that the checks would be for modified lines in a PR only, not for the whole codebase (for now!). I'll update the fist comment.

**Comment by techtonik** [comment]
For modified lines I guess `git diff` + `grep`.

**Comment by bilderbuchi** [comment]
> We just have to state that the line endings should always be linux-like in the repo, and it's up to the developer to correctly setup git (of course we can help with this part and document it).

The best way to do this is with a `.gitattributes` file (as I've already seen pointed out elsewhere) - this at the same time state your intent, and enforces it, so you don't need to rely on contributors figuring out a correct git config for their system.

**Comment by Nodd** [comment]
@bilderbuchi Yes, but currently the files in the repo have a mixed state, and brutally adding a .gitattributes would create a gigantic PR, so we're trying to take a softer route, i.e. fixing things little by little, then addins a .gitattributes.
If you have any ideas on how it could work easily do not hesitate to comment.

**Comment by bilderbuchi** [comment]
Basically, there is no easy way to do this. I have previously migrated another large-ish OSS project to gitattributes, also because users encountered weird line-ending problems pretty often. 
What I found worked best is to just do two commits (add the `.gitattributes`, then do the normalization). Yes, the changes are huge, and yes, they will confuse some open PRs (best to do this after a release, or whenever amount of open PRs is lowest), but then the work is done, and the changes are just whitespace (so you can ignore them by appending `?w=1` to all relevant Github URLs).

Your "softer" route will, in the end, just create more commit noise (if you plan to normalize line endings file-by-file over time), and more bad endings will creep in as long as no gitattributes is in place.

Be aware that there is a git-merge option for folding in the "upstream" normalization harmoniously, which could invalidate the point about the affected PRs.

**Comment by bilderbuchi** [comment]
> Be aware that there is a git-merge option for folding in the "upstream" normalization harmoniously, which could invalidate the point about the affected PRs.

Indeed, I found the command: `git merge master -s recursive -X renormalize`, see [here](http://git-scm.com/docs/git-merge) -> renormalize.

**Comment by ccordoba12** [comment]
@bilderbuchi, thanks for the advice.

@Nodd, let's wait to merge several big pull request before doing this then :-) I'll let you know when things are ready.

**Comment by goanpeca** [comment]
@ccordoba12 I think we will never get to the point where we have a low number of PRs so we will have to push this change sooner than later, even if it means messing some PRs and authors in the process.

I say we do it now, get over with this and move on...

**Comment by ccordoba12** [comment]
Ok, but let's merge first PR #2932. That's the biggest one right now and I think it'd give @blink1073 several headaches to try to update it :-)

That will force me also to find why that one is not working for me!!

**Comment by goanpeca** [comment]
:+1: we need a deadline and a counter :-po

**Comment by Nodd** [comment]
> let's wait to merge several big pull request before doing this then :-)

That's what we said last time :smiley:

**Comment by ccordoba12** [comment]
As i said, let's merge that @blink1073 PR and then we move to gitattributes ;-)

**Comment by Nodd** [comment]
There is a `--diff` option in pep8 that can be used:
```
  --diff               report only lines changed according to the unified diff
                       received on STDIN
```
(http://pep8.readthedocs.org/en/latest/intro.html)
